linker.ts

```diff
@@ -1,72 +1,133 @@
-      for (const lib in libraries[libraryName]) {
-        librariesComplete[lib] = libraries[libraryName][lib];
-        librariesComplete[libraryName + ':' + lib] = libraries[libraryName][lib];
-      }
-    } else {
-      // backwards compatible API for early solc-js versions
-      const parsed = libraryName.match(/^([^:]+):(.+)$/);
-      if (parsed) {
-        librariesComplete[parsed[2]] = libraries[libraryName];
+
+  for (const [libraryName, library] of Object.entries(libraries)) {
+    if (isNil(library)) continue;
+
+    // API compatible with the standard JSON i/o
+    if (isObject(library)) {
+      for (const [key, value] of Object.entries(library)) {
+        librariesComplete[key] = value;
+        librariesComplete[`${libraryName}:${key}`] = value;
       }
-      librariesComplete[libraryName] = libraries[libraryName];
+
+      continue;
+    }
+
+    // backwards compatible API for early solc-js versions
+    const parsed = libraryName.match(/^([^:]+):(.+)$/);
+
+    if (!isNil(parsed)) {
+      librariesComplete[parsed[2]] = library;
     }
+
+    librariesComplete[libraryName] = library;
   }
 
   for (const libraryName in librariesComplete) {
     let hexAddress = librariesComplete[libraryName];
-    if (hexAddress.slice(0, 2) !== '0x' || hexAddress.length > 42) {
-      throw new Error('Invalid address specified for ' + libraryName);
+
+    if (!hexAddress.startsWith('0x') || hexAddress.length > 42) {
+      throw new Error(`Invalid address specified for ${libraryName}`);
     }
+
     // remove 0x prefix
-    hexAddress = hexAddress.slice(2);
-    hexAddress = Array(40 - hexAddress.length + 1).join('0') + hexAddress;
-
-    // Support old (library name) and new (hash of library name)
-    // placeholders.
-    const replace = function (name) {
-      // truncate to 37 characters
-      const truncatedName = name.slice(0, 36);
-      const libLabel = '__' + truncatedName + Array(37 - truncatedName.length).join('_') + '__';
-      while (bytecode.indexOf(libLabel) >= 0) {
-        bytecode = bytecode.replace(libLabel, hexAddress);
-      }
-    };
+    hexAddress = hexAddress.slice(2).padStart(40, '0');
 
-    replace(libraryName);
-    replace(libraryHashPlaceholder(libraryName));
+    bytecode = replacePlaceholder(bytecode, libraryName, hexAddress);
+    bytecode = replacePlaceholder(bytecode, libraryHashPlaceholder(libraryName), hexAddress);
   }
 
   return bytecode;
-};
+}
 
-const findLinkReferences = function (bytecode) {
+/**
+ * As of Solidity 0.4.11 the compiler supports standard JSON input and output
+ * which outputs a _link references_ map. This gives a map of library names to
+ * offsets in the bytecode to replace the addresses at. It also doesn't have
+ * the limitation on library file and contract name lengths.
+ *
+ * There is a method available here which can find such link references in
+ * bytecode produced by an older compiler.
+ *
+ * @param {string} bytecode
+ * @returns {string} LinkedReferences
+ */
```

Comment:
Here's a more accurate piece of documentation you can use for this:

```suggestion
/**
 * Finds locations of all library address placeholders in the hex-encoded bytecode.
 * Returns information in a format matching `evm.bytecode.linkReferences` output
 * in Standard JSON.
 *
 * See [Library Linking](https://docs.soliditylang.org/en/latest/using-the-compiler.html#library-linking)
 * for a full explanation of library placeholders and linking process.
 *
 * WARNING: The output matches `evm.bytecode.linkReferences` exactly only in
 * case of old-style placeholders created from fully qualified library names of no more than 36
 * characters, and even then only if the name does not start or end with an underscore.
 * This is different from `evm.bytecode.linkReferences`, which uses fully qualified library
 * names. This is a limitation of the placeholder format - the fully qualified names are not preserved
 * in the compiled bytecode and cannot be reconstructed without external information.
 *
 * @param bytecode Hex-encoded bytecode string.
 *
 * @returns linkReferences A mapping between library labels and their locations in the
 *     bytecode. In case of old-style placeholders the label is a fully qualified library name
 *     truncated to 36 characters. For new-style placeholders it's the first 34 characters of the
 *     hex-encoded hash of the fully qualified library name, with a leading and trailing $ character added.
 *     Note that the offsets and lengths refer to the *binary* (not hex-encoded) bytecode, just
 *     like in `evm.bytecode.linkReferences`.
 */
```